audit/audit.go

```diff
@@ -0,0 +1,272 @@
+/*
+Package audit implements a smart client to a remote audit server. This handles any connection problems silently,
+while providing ways to detect problems through whatever alerting method is desired. It also handles an exponential
+backoff when trying to reconnect to the remote audit server as to prevent any overwhelming of the remote audit server.
+
+This is the preferred way to connect to a remote audit server. There is a more low-level way to connect to a remote
+audit server in the base package. This one allows for you to completely customize how you want to handle issues,
+but it is more work to do so.
+
+Example using the domainsocket package:
+
+	// Create a function that will create a new connection to the remote audit server.
+	// We use this function to create a new connection when the connection is broken.
+	cc := func() (conn.Audit, error) {
+		return conn.NewDomainSocket()
+	}
+
+	c, err := audit.New(cc)
+	if err != nil {
+		// Handle error.
+	}
+	defer c.Close(context.Background())
+
+	// This is optional if you want to get notifications of logging problems.
+	go func() {
+		for notifyMsg := range c.Notify() {
+			// Handle error notification.
+			// You can log them or whatever you want to do.
+		}
+	}()
+
+	// Send a message to the remote audit server.
+	if err := c.Send(context.Background(), msgs.Msg{<add record information>}); err != nil {
+		// Handle error.
+		// Errors here will either be:
+		// 1. base.ErrValidation , which means your message is invalid.
+		// 2. A standard error, which means there is an error condition we haven't categorized yet.
+	}
+*/
+package audit
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	"runtime"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/microsoft/go-otel-audit/audit/base"
+	"github.com/microsoft/go-otel-audit/audit/conn"
+	"github.com/microsoft/go-otel-audit/audit/msgs"
+
+	"github.com/Azure/retry/exponential"
+)
+
+// NotifyError is an error that is sent to the Notify channel when the connection to the remote audit server is broken.
+type NotifyError struct {
+	// Time is the time the error occurred.
+	Time time.Time
+	// Err is the error that happened. This will be a an base.ErrConnection or base.ErrQueueFull .
+	Err error
+}
+
+// Client is an audit server smart client. This handles any connection problems silently, while providing ways to
+// detect problems through whatever alerting method is desired.
+type Client struct {
+	// client holds the current audit client. If this is nil, it means we are trying to get a new client.
+	// When that occurs, we will do a noop on Send().
```

Comment:
Not queue the msg (or block or something?)

Note to myself to come back to this quesiton once I've read all the code.

Reply:
Ignore, this is a comment for a version of the code that doesn't exist.